handler.go

```diff
@@ -0,0 +1,214 @@
+package dhcp
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/insomniacslk/dhcp/dhcpv4"
+	"github.com/tinkerbell/dhcp/data"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// BackendReader is the interface that wraps the Read method.
+//
+// Backends implement this interface to provide DHCP data to the DHCP server.
+type BackendReader interface {
+	// Read data (from a backend) based on a mac address
+	// and return DHCP headers and options, including netboot info.
+	Read(context.Context, net.HardwareAddr) (*data.Dhcp, *data.Netboot, error)
+}
+
+// handleFunc is the main handler for DHCPv4 packets.
+func (s *Server) handleFunc(conn net.PacketConn, peer net.Addr, m *dhcpv4.DHCPv4) {
+	log := s.Log.WithValues("mac", m.ClientHWAddr.String())
+	tracer := otel.Tracer("DHCP")
+	ctx, span := tracer.Start(s.ctx, "DHCP Packet Received",
+		trace.WithAttributes(attribute.String("MAC", m.ClientHWAddr.String())),
+		trace.WithAttributes(attribute.String("MessageType", m.MessageType().String())),
+	)
+	defer span.End()
+	var reply *dhcpv4.DHCPv4
+	switch mt := m.MessageType(); mt {
+	case dhcpv4.MessageTypeDiscover:
+		reply = s.handleDiscover(ctx, tracer, m)
+	case dhcpv4.MessageTypeRequest:
+		reply = s.handleRequest(ctx, tracer, m)
+	case dhcpv4.MessageTypeRelease:
+		s.handleRelease(ctx, m)
+	default:
+		log.Info("received unknown message type", "type", mt)
+	}
+	c := codes.Ok
+	cmsg := fmt.Sprintf("no reply send DHCP: %v", m.MessageType())
+	if reply != nil {
```

Comment:
I don't see logic updating the span to reflect that the reply was unexpectedly nil.
That seems like a case we might care about.

Reply:
in boots I handled this by calling span.SetStatus in the error handling path